client/consensus/grandpa/authorities.go

```diff
@@ -0,0 +1,814 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"golang.org/x/exp/constraints"
+	"golang.org/x/exp/slices"
+)
+
+var (
+	errInvalidAuthoritySet                           = errors.New("invalid authority set, either empty or with an authority weight set to 0")
+	errDuplicateAuthoritySetChanges                  = errors.New("duplicate authority set hashNumber")
+	errMultiplePendingForcedAuthoritySetChanges      = errors.New("multiple pending forced authority set changes are not allowed")
+	errForcedAuthoritySetChangeDependencyUnsatisfied = errors.New("a pending forced authority set hashNumber could not be applied since it must be applied after the pending standard hashNumber")
+	errForkTree                                      = errors.New("invalid operation in the pending hashNumber tree")
+)
+
+// SharedAuthoritySet A shared authority set
+type SharedAuthoritySet[H comparable, N constraints.Unsigned] struct {
+	mtx   sync.Mutex
+	inner AuthoritySet[H, N]
+}
+
+// delayedKinds Kinds of delays for pending changes.
+type delayedKinds[N constraints.Unsigned] interface {
+	Finalized | Best[N]
+}
+
+// delayKind struct to represent delayedKinds
+type delayKind struct {
+	value interface{}
+}
+
+func setDelayKind[N constraints.Unsigned, T delayedKinds[N]](delayKind *delayKind, val T) {
+	delayKind.value = val
+}
+
+func newDelayKind[N constraints.Unsigned, T delayedKinds[N]](val T) delayKind {
+	delayKind := delayKind{}
+	setDelayKind[N](&delayKind, val)
+	return delayKind
+}
+
+// Finalized Depth in finalized chain.
+type Finalized struct{}
+
+// Best Depth in best chain. The median last finalized block is calculated at the time the
+// hashNumber was signaled.
+type Best[N constraints.Unsigned] struct {
+	medianLastFinalized N
+}
+
+// IsDescendentOf is the function definition to determine if target is a descendant of base
+type IsDescendentOf[H comparable] func(base, target H) (bool, error)
+
+// setIDNumber represents the set id and block number of an authority set hashNumber
+type setIDNumber[N constraints.Unsigned] struct {
+	setId       uint64
+	blockNumber N
+}
+
+// generic representation of hash and number tuple
+type hashNumber[H, N any] struct {
+	hash   H
+	number N
+}
+
+// key is used to represent a tuple ordered first by effective number and then by signal-block number
+type key[N any] struct {
+	effectiveNumber   N
+	signalBlockNumber N
+}
```

Comment:
Since this is only used in `AuthoritySet.addForcedChange` move this inside that functiuon.